Feat/crypto setup

magsteen opened the pull request:
Added functions for easily generating keypair(s) and encrypt/decrypt byte vectors. 

Added one unit test for encryption/decryption.
Dropped test for keygeneration as it is mostly handled by an external create and our function will merely return other types in a wrapper. 

What do you guys think? Is the encryption valid? Anything more we should add to crypto.rs?

![image](https://user-images.githubusercontent.com/73636441/158659533-f5eac324-4914-4b9e-99b7-32086d0c1f30.png)

magsteen committed:
Fixed project structure to allow dependencies

magsteen committed:
Created utillity funcs for  cryptography

magsteen committed:
Merge branch 'master' into feat/crypto-setup

ngorke commented on core/src/crypto.rs:
```diff
@@ -0,0 +1,66 @@
+}
+
+fn create_keys(n: u8) -> Vec<KeyPair> {
+    let mut keys = Vec::new();
```
You can use Vec::with_capacity here to pre-allocate.

magsteen replied:
Good point!

ngorke commented on core/src/crypto.rs:
```diff
@@ -0,0 +1,66 @@
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_create_key() {}
+
+    #[test]
+    fn test_create_keys() {}
```
You can remove the functions if you aren't going to have the tests

magsteen replied:
Yes, I agree. Was wondering if we should test this somehow?

ngorke replied:
Yes, but I don't think it will provide much value.

ngorke commented on core/src/crypto.rs:
```diff
@@ -0,0 +1,66 @@
+
+    #[test]
+    fn test_xor() {
+        let mut pair = create_key();
+
+        let msg = "Hello".to_owned().into_bytes();
+
+        let encrypted_msg = pair.public_key.xor(&msg);
+        let unencrypted_msg = pair.public_key.xor(&encrypted_msg);
+
+        assert_eq!(msg, unencrypted_msg);
```
You should rather test for an expected output, not if the two invocations cancel each other.

magsteen replied:
Yes :point_right:

ngorke commented on core/src/crypto.rs:
```diff
@@ -0,0 +1,66 @@
+    keys
+}
+
+trait Key {
+    fn xor(&mut self, data: &Vec<u8>) -> Vec<u8>;
+}
+
+macro_rules! impl_key {
+    ($($t:ty),+) => {
+        $(impl Key for $t {
+            fn xor(&mut self, data: &Vec<u8>) -> Vec<u8> {
+                let key_bytes = self.as_bytes();
+                let key_len = key_bytes.len();
+                let mut res: Vec<u8> = Vec::new();
+                for i in 0..data.len() {
+                    res.push(data[i] ^ key_bytes[i % key_len])
+                }
+                return res
+            }
+        })*
+    }
+}
+
+impl_key!(SharedSecret, PublicKey);
```
Why do we need xor?

magsteen replied:
Well I do realize now that it is useless for the public key, but still useful for encrypting/decrypting with a symmetric key.

ngorke replied:
We cannot simply use XOR for encryption/decryption, as it is not safe to use it for symmetric encryption when we reuse the encryption key.

ngorke commented on core/src/crypto.rs:
```diff
@@ -0,0 +1,66 @@
+use rand_core::OsRng;
+use x25519_dalek::{EphemeralSecret, PublicKey, SharedSecret};
+
+struct KeyPair {
+    secret: EphemeralSecret,
+    public_key: PublicKey,
+}
```
This is not a key pair.

magsteen replied:
What do you think we should call it?

ngorke replied:
Secret :feelsgood:

ngorke commented on core/src/crypto.rs:
```diff
@@ -0,0 +1,66 @@
+    public_key: PublicKey,
+}
+
+fn create_key() -> KeyPair {
+    let secret = EphemeralSecret::new(OsRng);
+    let public_key = PublicKey::from(&secret);
+    KeyPair { secret, public_key }
+}
```
Also, this should be a new() method in the KeyPair struct.

magsteen committed:
Refactored with encryption from a library

magsteen committed:
Merge branch 'feat/crypto-setup' of github.com:tomrsae/ronion into feat/crypto-setup

magsteen committed:
Removed unfinished

magsteen commented:
Refactored and added a AES library we can use. We only have to generate ciphers and use the inbuilt encrypt/decrypt functions. The library provides a good variety of functions i think. :)

magsteen committed:
Refactor visibility

ngorke commented on core/src/lib.rs:
```diff
@@ -1 +1,2 @@
-// lib code
\ No newline at end of file
+// lib code
```
plz no

magsteen committed:
Removed comments

magsteen committed:
fix lib.rs

ngorke commented on core/Cargo.toml:
```diff
@@ -9,3 +9,6 @@ Library for using the onion protocol to securely and anonymously connect to serv
 repository = "https://github.com/tomrsae/ronion"
 
 [dependencies]
+x25519-dalek = "2.0.0-pre.1"
```
🍣 -hi

ngorke commented on core/src/crypto.rs:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    pub fn create_secrets(n: usize) -> Vec<Secret> {
+        let mut secrets = Vec::with_capacity(n);
+        for _ in 0..n {
+            secrets.push(Secret::new())
+        }
+        secrets
```
You could also do
`
(0..n).iter().map(|_| Secret::new()).collect()
`

ngorke commented on core/src/crypto.rs:
```diff
@@ -0,0 +1,40 @@
+
+    pub fn gen_cipher(&mut self) -> Aes256 {
+        let key = self.gen_pub_key();
+        match Aes256::new_from_slice(key.as_bytes()) {
```
You could also just call .expect on the result

ngorke reviewed (changes requested):
🍣 -hi

magsteen committed:
PR refactoring/optimization

tomrsae commented:
Where are the tests? I can see the screenshot of it, but they're never ran, nor can I find the code for it 🤔

ngorke reviewed (approved):
Nice r🧅!
